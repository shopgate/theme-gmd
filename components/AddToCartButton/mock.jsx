/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
/**
 * Mocked AddToCartButton.
 * @type {MockedAddToCartButton}
 */
// eslint-disable-next-line react/prefer-stateless-function
export const MockedAddToCartButton = class extends Component {
  static propTypes = {
    handleAddToCart: PropTypes.func,
  };
  static defaultProps = {
    handleAddToCart: () => {},
  };
  /**
   * Renders mocked button.
   * @return {JSX}
   */
  render() {
    return (
      <button onClick={this.props.handleAddToCart} />
    );
  }
};
