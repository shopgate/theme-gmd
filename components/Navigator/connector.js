import { connect } from 'react-redux';
import { navigate } from '@shopgate/pwa-common/action-creators/router';
import fetchSearchSuggestions from '@shopgate/pwa-common-commerce/search/actions/fetchSearchSuggestions';
import {
  getBackgroundColor,
  isFilterOpen,
  isEnabled,
  isNavSearchFieldActive,
  isSearchShowing,
  isTitleShowing,
  showLoadingBar,
  getTextColor,
} from './selectors';

/**
 * @param {Object} state The current application state.
 * @return {Object}
 */
const mapStateToProps = state => ({
  backgroundColor: getBackgroundColor(state),
  filterOpen: isFilterOpen(state),
  navigatorEnabled: isEnabled(state),
  searchActive: isNavSearchFieldActive(state),
  showSearch: isSearchShowing(state),
  showTitle: isTitleShowing(state),
  showLoadingBar: showLoadingBar(state),
  textColor: getTextColor(state),
});

/**
 * @param {Function} dispatch The store dispatcher.
 * @return {Object}
 */
const mapDispatchToProps = dispatch => ({
  navigate: (action, pathname) => dispatch(navigate({
    action,
    pathname,
  })),
  fetchSearchSuggestions: searchPhrase => dispatch(fetchSearchSuggestions(searchPhrase)),
});

export default connect(mapStateToProps, mapDispatchToProps);
