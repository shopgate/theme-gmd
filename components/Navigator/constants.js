/**
 * Copyright (c) 2017-present, Shopgate, Inc. All rights reserved.
 *
 * This source code is licensed under the Apache 2.0 license found in the
 * LICENSE file in the root directory of this source tree.
 */

 /**
  * Action Types.
  */

export const TOGGLE_NAVIGATOR_CART_ICON = 'TOGGLE_NAVIGATOR_CART_ICON';
export const TOGGLE_NAVIGATOR_SEARCH = 'TOGGLE_NAVIGATOR_SEARCH';
export const TOGGLE_NAV_DRAWER = 'TOGGLE_NAV_DRAWER';
export const TOGGLE_PROGRESS_BAR = 'TOGGLE_PROGRESS_BAR';
export const TOGGLE_LOGIN = 'TOGGLE_LOGIN';
export const SET_NAVIGATOR_BACKGROUND = 'SET_NAVIGATOR_BACKGROUND';
export const SET_NAVIGATOR_COLOR = 'SET_NAVIGATOR_COLOR';
export const SET_NAVIGATOR_SEARCH_QUERY = 'SET_NAVIGATOR_SEARCH_QUERY';
export const SET_FILTER_OPENED = 'SET_FILTER_OPENED';
export const SET_FILTER_CLOSED = 'SET_FILTER_CLOSED';
export const SET_FILTER_ATTRIBUTE_OPENED = 'SET_FILTER_ATTRIBUTE_OPENED';
export const SET_FILTER_ATTRIBUTE_CLOSED = 'SET_FILTER_ATTRIBUTE_CLOSED';
export const SET_SEARCH_ENABLED = 'SET_SEARCH_ENABLED';
export const SET_SEARCH_DISABLED = 'SET_SEARCH_DISABLED';
export const SET_ICON_SHADOW_ENABLED = 'SET_ICON_SHADOW_ENABLED';
export const SET_ICON_SHADOW_DISABLED = 'SET_ICON_SHADOW_DISABLED';
export const SET_TITLE_ENABLED = 'SET_TITLE_ENABLED';
export const SET_TITLE_DISABLED = 'SET_TITLE_DISABLED';
export const SET_NAVIGATOR_DISABLED = 'SET_NAVIGATOR_DISABLED';
export const SET_NAVIGATOR_ENABLED = 'SET_NAVIGATOR_ENABLED';
