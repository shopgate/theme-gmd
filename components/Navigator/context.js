import React from 'react';

export const NavigatorContext = React.createContext();
