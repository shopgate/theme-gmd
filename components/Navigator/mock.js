import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const router = {
  routing: false,
  stack: [
    {
      id: '44a34767-2683-4414-9de4-fd736808fdab',
      params: {},
      pathname: '/',
      pattern: '/',
      query: {},
      state: {},
    },
  ],
};

export const defaultState = {
  navigator: {
    backgroundColor: 'red',
    enabled: true,
    filterOpen: false,
    filterAttributeOpen: false,
    loginOpen: false,
    searchActive: false,
    showCartIcon: true,
    showIconShadow: false,
    showProgressBar: true,
    showSearch: true,
    showTitle: true,
    textColor: 'white',
    viewTracking: false,
  },
  search: {
    suggestion: {},
  },
  router,
  view: {
    isLoading: false,
  },
  ui: {
    general: {
      title: null,
    },
  },
};

/**
 * Creates store.
 * @param {Object} props Props which would override state.navigator props.
 * @return {Object}
 */
export const getStore = props => configureStore([thunk])({
  ...defaultState,
  navigator: {
    ...defaultState.navigator,
    ...props,
  },
  search: {
    suggestions: {},
  },
});
