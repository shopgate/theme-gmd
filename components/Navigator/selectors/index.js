import { createSelector } from 'reselect';
import {
  getHistoryPathname,
  getQueryParamsAsString,
} from '@shopgate/pwa-common/selectors/history';
import {
  getCartProductCount,
  getProductPendingCount,
} from '@shopgate/pwa-common-commerce/cart/selectors';
import { isViewLoading } from '@shopgate/pwa-common/selectors/view';

/**
 * Selects the navigator state.
 * @param {Object} state The global state.
 * @return {Object}
 */
export const getNavigatorState = state => state.navigator;

/**
 * Selects the current background color from navigator state.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getBackgroundColor = createSelector(
  getNavigatorState,
  state => state.backgroundColor
);

/**
 * @param {Object} state The global state.
 * @return {Object}
 */
export const isFilterOpen = createSelector(
  getNavigatorState,
  state => state.filterOpen
);

/**
 * Selects the current text color from navigator state.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getTextColor = createSelector(
  getNavigatorState,
  state => state.textColor
);

/**
 * Selects the current state of shadow of the navigator's icons.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isIconShadowShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showIconShadow
);

/**
 * Selects the current state from navigator search.
 * @param {Object} state The global state.
 * @return {string}
 */
export const getSearchPhrase = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.searchPhrase
);

/**
 * Selects the current state from navigator search.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isSearchShowing = createSelector(
  getNavigatorState,
  state => state.showSearch
);

/**
 * Selects the current state from navigator progress bar.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isProgressBarShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showProgressBar
);

/**
 * @param {Object} state The global state.
 * @return {boolean}
 */
export const showLoadingBar = createSelector(
  isProgressBarShowing,
  isViewLoading,
  (progressBarShowing, viewLoading) => !!(progressBarShowing && viewLoading)
);

/**
 * Selects the current state from navigator title.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isTitleShowing = createSelector(
  getNavigatorState,
  state => state.showTitle
);

/**
 * Selects the current state state from navigator disabled status.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isEnabled = createSelector(
  getNavigatorState,
  state => state.enabled
);

/**
 * Selects the current state state from navigator disabled status.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isLoadingBarShowing = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.showLoadingBar
);

/**
 * Selects the current state of the navigation drawer.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isNavDrawerActive = createSelector(
  getNavigatorState,
  navigatorState => navigatorState.navDrawerActive
);

/**
 * Selects the current state of the navigation drawer.
 * @param {Object} state The global state.
 * @return {string}
 */
export const isNavSearchFieldActive = createSelector(
  getNavigatorState,
  state => state.searchActive
);

/**
 * Checks if the cart button is available.
 * @return {boolean}
 */
export const isCartButtonVisible = createSelector(
  getCartProductCount,
  getProductPendingCount,
  getNavigatorState,
  (count, pendingCount, navigator) => (
    navigator.showCartIcon && (count + pendingCount) > 0
  )
);

/**
 * Selects the navigator state for tracking.
 * @param {Object} state The application state.
 * @return {Object}
 */
export const getTrackingNavigatorState = createSelector(
  getHistoryPathname,
  getQueryParamsAsString,
  (pathname, queryString) => ({
    pathname,
    queryString,
  })
);
