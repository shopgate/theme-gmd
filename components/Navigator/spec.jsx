import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';
import { ACTION_PUSH, ACTION_REPLACE } from '@virtuous/conductor/constants';
import mockRenderOptions from '@shopgate/pwa-common/helpers/mocks/mockRenderOptions';
import { NAVIGATE } from '@shopgate/pwa-common/constants/ActionTypes';
import { mockedPipelineRequestFactory } from '@shopgate/pwa-core/classes/PipelineRequest/mock';
import { getStore } from './mock';
import Navigator from './index';

jest.mock('./context');

// eslint-disable-next-line react/prop-types
jest.mock('./components/CartButton', () => ({ children }) => (<div>{children} </div>));
// eslint-disable-next-line react/prop-types
jest.mock('./components/NavButton', () => ({ children }) => (<div>{children} </div>));

let mockedRoutePattern = '/';
jest.mock('@virtuous/conductor-helpers/getCurrentRoute', () => () => ({ pattern: mockedRoutePattern }));

const mockedResolver = jest.fn();
jest.mock('@shopgate/pwa-core/classes/PipelineRequest', () => mockedPipelineRequestFactory((mockInstance, resolve, reject) => {
  mockedResolver(mockInstance, resolve, reject);
}));

const results = [
  [
    {
      type: NAVIGATE,
      action: ACTION_REPLACE,
      location: '/search?s=shirt',
      state: undefined,
    },
  ],
  [
    {
      type: NAVIGATE,
      action: ACTION_PUSH,
      location: '/search?s=skirt',
      state: undefined,
    },
  ],
];

describe('Navigator', () => {
  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should render nothing', () => {
    const component = mount(
      <Provider store={getStore({ enabled: false })}>
        <Navigator />
      </Provider>,
      mockRenderOptions
    );
    expect(component.html()).toBe(null);
  });
  it('should render the component and update when props change', () => {
    const component = mount(
      <Provider store={getStore({ enabled: true })}>
        <Navigator />
      </Provider>,
      mockRenderOptions
    );
    expect(component.find('header').exists()).toBe(true);
  });

  it('should render search', () => {
    const store = getStore({ enabled: true, showTitle: true, showSearch: true });

    const component = mount(
      <Provider store={store}>
        <Navigator />
      </Provider>,
      mockRenderOptions
    );

    expect(component.find('NavigatorSearch').exists()).toBe(true);
    expect(component).toMatchSnapshot();
  });

  it('setRoutePattern', () => {
    const component = mount(
      <Provider store={getStore({})}>
        <Navigator />
      </Provider>,
      mockRenderOptions
    );

    expect(component.find('Navigator').instance().state.routePattern).toEqual('');

    mockedRoutePattern = '/search';
    component.find('Navigator').instance().setRoutePattern();

    component.update();
    expect(component.find('Navigator').instance().state.routePattern).toEqual(mockedRoutePattern);
  });

  it('setSearchQuery', () => {
    const component = mount(
      <Provider store={getStore({})}>
        <Navigator />
      </Provider>,
      mockRenderOptions
    );

    component.find('Navigator').instance().setSearchQuery('skirt');
    expect(component.find('Navigator').instance().state.searchQuery).toEqual('skirt');
  });

  describe('search navigation', () => {
    let store;

    /* eslint-disable-next-line require-jsdoc */
    const createComponent = () => {
      store = getStore({ enabled: true, showTitle: true, showSearch: true });
      const component = mount(
        <Provider store={store}>
          <Navigator />
        </Provider>,
        mockRenderOptions
      );
      return component.find('Navigator');
    };

    afterEach(() => {
      store.clearActions();
    });

    it('should not navigate', () => {
      const active = false;
      const NavigatorWrapper = createComponent();
      NavigatorWrapper.instance().setState({
        searchQuery: 'shirt',
        routePattern: '/search',
        searchField: active,
      });

      NavigatorWrapper.instance().toggleSearchField(active);

      expect(store.getActions().length).toEqual(0);
    });

    it('should navigate search (replace)', () => {
      const NavigatorWrapper = createComponent();
      const replaceFromPath = '/search';

      NavigatorWrapper.instance().setState({
        searchQuery: 'shirt',
        routePattern: replaceFromPath,
        searchField: true,
      });
      NavigatorWrapper.instance().toggleSearchField(false, true);

      expect(store.getActions()).toEqual(results[0]);
    });

    it('should navigate search (push)', () => {
      const NavigatorWrapper = createComponent();
      const pushFromPath = '/category';

      NavigatorWrapper.instance().setState({
        searchQuery: 'skirt',
        routePattern: pushFromPath,
        searchField: true,
      });
      NavigatorWrapper.instance().toggleSearchField(false, true);

      expect(store.getActions()).toEqual(results[1]);
    });
  });
});
