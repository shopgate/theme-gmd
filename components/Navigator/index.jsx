import React, { PureComponent } from 'react';
import PropTypes from 'prop-types';
import throttle from 'lodash/throttle';
import Grid from '@shopgate/pwa-common/components/Grid';
import Portal from '@shopgate/pwa-common/components/Portal';
import * as portals from '@shopgate/pwa-common/constants/Portals';
import { SEARCH_PATH } from '@shopgate/pwa-common-commerce/search/constants';
import ProgressBar from '@shopgate/pwa-ui-shared/ProgressBar';
import { ACTION_PUSH, ACTION_REPLACE } from '@virtuous/conductor/constants';
import * as events from '@virtuous/conductor-events';
import getCurrentRoute from '@virtuous/conductor-helpers/getCurrentRoute';
import colors from 'Styles/colors';
import connect from './connector';
import NavButton from './components/NavButton';
import SearchButton from './components/SearchButton';
import CartButton from './components/CartButton';
import ApplyFilterButton from './components/ApplyFilterButton';
import Content from './components/Content';
import { NavigatorContext } from './context';
import styles from './style';

/**
 * The Navigator component.
 * @param {Object} props The component props.
 */
class Navigator extends PureComponent {
  static propTypes = {
    fetchSearchSuggestions: PropTypes.func.isRequired,
    navigate: PropTypes.func.isRequired,
    backgroundColor: PropTypes.string,
    filterOpen: PropTypes.bool,
    navigatorEnabled: PropTypes.bool,
    showLoadingBar: PropTypes.bool,
    showSearch: PropTypes.bool,
    showTitle: PropTypes.bool,
    textColor: PropTypes.string,
  };

  static defaultProps = {
    backgroundColor: colors.light,
    filterOpen: false,
    navigatorEnabled: true,
    showLoadingBar: false,
    showSearch: true,
    showTitle: true,
    textColor: colors.dark,
  };

  /**
   * @param {Object} props The component props.
   */
  constructor(props) {
    super(props);

    this.state = {
      routePattern: '',
      searchField: false,
      searchQuery: '',
    };

    events.onDidPush(this.setRoutePattern);
    events.onDidPop(this.setRoutePattern);
    events.onDidReplace(this.setRoutePattern);
    events.onDidReset(this.setRoutePattern);
  }

  /**
   * Initially set the route pattern. This is necessary to determine which
   * route is the first when the app is opened.
   */
  componentDidMount() {
    this.setRoutePattern();
  }

  /**
   * @param {string} id The id of the route that entered.
   */
  setRoutePattern = () => {
    const { pattern } = getCurrentRoute();

    if (this.state.routePattern !== pattern) {
      this.setState({
        routePattern: pattern,
      });
    }
  }

  /**
   * @param {string} query The new search query.
   */
  setSearchQuery = (query) => {
    if (this.state.searchQuery !== query) {
      this.setState({
        searchQuery: query,
      }, this.fetchSuggestions);
    }
  }

  /**
   * Fetch search suggestions with the locally set search query.
   */
  fetchSuggestions = throttle(() => {
    this.props.fetchSearchSuggestions(this.state.searchQuery);
  }, 1000)

  /**
   * @param {boolean} active The new state for the search field.
   * @param {boolean} submit Whether or not to submit the search.
   */
  toggleSearchField = (active = false, submit = false) => {
    if (this.state.searchField === active) {
      return;
    }

    this.setState({
      searchField: active,
    });

    if (submit && !active) {
      this.handleSubmitSearch();
    }
  }

  /**
   * Navigate to the search route. If we are already
   * on the search route then perform a history replace
   * action instead. This is to not pollute the history.
   */
  handleSubmitSearch = () => {
    if (this.state.searchQuery) {
      const isSearchRoute = (this.state.routePattern === SEARCH_PATH);
      const action = (isSearchRoute) ? ACTION_REPLACE : ACTION_PUSH;

      this.props.navigate(action, `/search?s=${this.state.searchQuery}`);
    }
  }

  /**
   * Renders the component.
   * @returns {JSX}
   */
  render() {
    if (!this.props.navigatorEnabled) {
      return null;
    }

    const headerStyle = {
      background: this.props.backgroundColor,
      color: this.props.textColor,
    };

    const context = {
      ...this.state,
      toggleSearchField: this.toggleSearchField,
      setSearchQuery: this.setSearchQuery,
    };

    return (
      <NavigatorContext.Provider value={context}>
        <Portal name={portals.NAV_BAR_BEFORE} />
        <Portal name={portals.NAV_BAR}>
          <header
            className={styles.header}
            data-test-id="Navigator"
            role="banner"
            style={headerStyle}
          >
            <Portal name={portals.NAV_BAR_NAVIGATOR_BEFORE} />
            <Portal name={portals.NAV_BAR_NAVIGATOR}>
              <Grid className={styles.grid} component="section" wrap={false}>
                <Portal name={portals.NAV_BAR_NAVIGATOR_NAV_BUTTON_BEFORE} />
                <Portal name={portals.NAV_BAR_NAVIGATOR_NAV_BUTTON} props={{ NavButton }}>
                  <NavButton pattern={this.state.routePattern} />
                </Portal>
                <Portal name={portals.NAV_BAR_NAVIGATOR_NAV_BUTTON_AFTER} />
                <Portal name={portals.NAV_BAR_NAVIGATOR_CENTER_BEFORE} />
                <Portal name={portals.NAV_BAR_NAVIGATOR_CENTER}>
                  <Grid.Item className={styles.title} component="div" grow={1}>
                    {this.props.showTitle &&
                      <Content routePattern={this.state.routePattern} />
                    }
                  </Grid.Item>
                </Portal>
                <Portal name={portals.NAV_BAR_NAVIGATOR_CENTER_AFTER} />
                <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS_BEFORE} />
                <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS} >
                  {(this.props.filterOpen) &&
                    <div className={styles.applyButton}>
                      <ApplyFilterButton />
                    </div>
                  }
                  {this.props.showSearch &&
                    <SearchButton />
                  }
                  <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS_CART_BUTTON_BEFORE} />
                  <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS_CART_BUTTON} props={{ CartButton }}>
                    <CartButton />
                  </Portal>
                  <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS_CART_BUTTON_AFTER} />
                </Portal>
                <Portal name={portals.NAV_BAR_NAVIGATOR_ICONS_AFTER} />
              </Grid>
            </Portal>
            <Portal name={portals.NAV_BAR_NAVIGATOR_AFTER} />
            <Portal name={portals.NAV_BAR_PROGRESS_BAR_BEFORE} />
            <Portal name={portals.NAV_BAR_PROGRESS_BAR} props={{ ProgressBar }}>
              <ProgressBar isVisible={this.props.showLoadingBar} />
            </Portal>
            <Portal name={portals.NAV_BAR_PROGRESS_BAR_AFTER} />
          </header>
        </Portal>
        <Portal name={portals.NAV_BAR_AFTER} />
      </NavigatorContext.Provider>
    );
  }
}

export default connect(Navigator);
