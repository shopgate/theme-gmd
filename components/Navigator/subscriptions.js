import { searchDidEnter$ } from '@shopgate/pwa-common-commerce/search/streams';
import { cartDidEnter$, cartDidLeave$ } from '@shopgate/pwa-common-commerce/cart/streams';
import { filterDidEnter$, filterDidLeave$ } from '@shopgate/pwa-common-commerce/filter/streams';
import { routeDidChange$ } from '@shopgate/pwa-common/streams';
import { getSearchPhrase } from '@shopgate/pwa-common/selectors/history';
import toggleNavSearchField from 'Components/Navigator/actions/toggleNavSearchField';
import { setSearchPhrase } from './action-creators';
import enableNavigatorSearch from './actions/enableNavigatorSearch';
import disableNavigatorSearch from './actions/disableNavigatorSearch';
import toggleCartIcon from './actions/toggleCartIcon';

// Derived streams.
export const cartFilterRoutesDidEnter$ = cartDidEnter$.merge(filterDidEnter$);
export const cartFilterRoutesDidLeave$ = cartDidLeave$.merge(filterDidLeave$);

/**
 * Navigator subscriptions.
 * @param {Function} subscribe The subscribe function.
 */
export default function navigator(subscribe) {
  /**
   * Gets triggered on all route changes.
   */
  subscribe(routeDidChange$, ({ dispatch }) => {
    dispatch(toggleNavSearchField(false));
  });

  /**
   * Gets triggered on entering the search route.
   */
  subscribe(searchDidEnter$, ({ dispatch, getState }) => {
    const state = getState();

    // If search input is empty, set it to the value of the search query param.
    if (!state.navigator.searchPhrase) {
      dispatch(setSearchPhrase(getSearchPhrase(state)));
    }
  });

  /**
   * Gets triggered when leaving the cart or filter route.
   */
  subscribe(cartFilterRoutesDidLeave$, ({ dispatch }) => {
    dispatch(toggleCartIcon(true));
    dispatch(enableNavigatorSearch());
  });

  /**
   * Gets triggered when entering the cart or filter route.
   */
  subscribe(cartFilterRoutesDidEnter$, ({ dispatch }) => {
    dispatch(toggleCartIcon(false));
    dispatch(disableNavigatorSearch());
  });
}
