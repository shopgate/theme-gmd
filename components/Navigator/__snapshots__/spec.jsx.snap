// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigator should render search 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <Connect(Navigator)>
    <Navigator
      backgroundColor="red"
      fetchSearchSuggestions={[Function]}
      filterOpen={false}
      navigate={[Function]}
      navigatorEnabled={true}
      searchActive={false}
      showLoadingBar={false}
      showSearch={true}
      showTitle={true}
      textColor="white"
    >
      <Provider
        value={
          Object {
            "routePattern": "",
            "searchField": false,
            "searchQuery": "",
            "setSearchQuery": [Function],
            "toggleSearchField": [Function],
          }
        }
      >
        <div
          value={
            Object {
              "routePattern": "",
              "searchField": false,
              "searchQuery": "",
              "setSearchQuery": [Function],
              "toggleSearchField": [Function],
            }
          }
        >
          <Portal
            name="nav-bar.before"
            props={null}
          />
          <Portal
            name="nav-bar"
            props={null}
          >
            <header
              className="css-1ul2v41"
              data-test-id="Navigator"
              role="banner"
              style={
                Object {
                  "background": "red",
                  "color": "white",
                }
              }
            >
              <Portal
                name="nav-bar.navigator.before"
                props={null}
              />
              <Portal
                name="nav-bar.navigator"
                props={null}
              >
                <Grid
                  className="css-6q3zgq"
                  component="section"
                  wrap={false}
                >
                  <section
                    className="css-6q3zgq css-h7r55a"
                  >
                    <Portal
                      name="nav-bar.navigator.nav-button.before"
                      props={null}
                    />
                    <Portal
                      name="nav-bar.navigator.nav-button"
                      props={
                        Object {
                          "NavButton": [Function],
                        }
                      }
                    >
                      <Component>
                        <div>
                           
                        </div>
                      </Component>
                    </Portal>
                    <Portal
                      name="nav-bar.navigator.nav-button.after"
                      props={null}
                    />
                    <Portal
                      name="nav-bar.navigator.center.before"
                      props={null}
                    />
                    <Portal
                      name="nav-bar.navigator.center"
                      props={null}
                    >
                      <GridItem
                        className="css-1kjnwv2"
                        component="div"
                        grow={1}
                        shrink={1}
                      >
                        <div
                          className="css-1kjnwv2 css-1gs0ko2"
                        >
                          <NavigatorContent
                            routePattern=""
                          >
                            <Connect(Title)>
                              <Title
                                dispatch={[Function]}
                                onClick={[Function]}
                                title=""
                              >
                                <div
                                  aria-hidden={true}
                                  onClick={[Function]}
                                >
                                  <div
                                    className="css-iykibm "
                                  >
                                    <Translate
                                      className=""
                                      params={Object {}}
                                      string=""
                                    />
                                  </div>
                                  <div
                                    className="css-iykibm "
                                    data-test-id="title: "
                                  >
                                    <Translate
                                      className=""
                                      params={Object {}}
                                      string=""
                                    />
                                  </div>
                                </div>
                              </Title>
                            </Connect(Title)>
                            <_default>
                              <Consumer>
                                <NavigatorSearch
                                  active={false}
                                  query=""
                                  setSearchQuery={[Function]}
                                  toggleSearchField={[Function]}
                                >
                                  <Transition
                                    appear={false}
                                    enter={true}
                                    exit={true}
                                    in={false}
                                    mountOnEnter={false}
                                    onEnter={[Function]}
                                    onEntered={[Function]}
                                    onEntering={[Function]}
                                    onExit={[Function]}
                                    onExited={[Function]}
                                    onExiting={[Function]}
                                    timeout={0}
                                    unmountOnExit={false}
                                  >
                                    <div
                                      className="css-24kq4i"
                                      style={
                                        Object {
                                          "transform": "translateY(-100%)",
                                        }
                                      }
                                    >
                                      <form
                                        className="css-sa13d4"
                                        data-test-id="Search"
                                        onSubmit={[Function]}
                                      >
                                        <input
                                          className="css-13u5ou6"
                                          onChange={[Function]}
                                          onFocus={[Function]}
                                          placeholder="search.placeholder"
                                          type="search"
                                          value=""
                                        />
                                      </form>
                                      <div
                                        aria-hidden={true}
                                        className="css-16d1k2p"
                                        onClick={[Function]}
                                        role="button"
                                      />
                                    </div>
                                  </Transition>
                                </NavigatorSearch>
                              </Consumer>
                            </_default>
                            <Component>
                              <Consumer>
                                <Suggestions
                                  isVisible={false}
                                />
                              </Consumer>
                            </Component>
                          </NavigatorContent>
                        </div>
                      </GridItem>
                    </Portal>
                    <Portal
                      name="nav-bar.navigator.center.after"
                      props={null}
                    />
                    <Portal
                      name="nav-bar.navigator.icons.before"
                      props={null}
                    />
                    <Portal
                      name="nav-bar.navigator.icons"
                      props={null}
                    >
                      <_default>
                        <Consumer>
                          <SearchButton
                            searchActive={false}
                            toggleSearchField={[Function]}
                          >
                            <button
                              className="css-1nlaj2c"
                              data-test-id="SearchButton"
                            >
                              <Ripple
                                className="css-ugy6xl"
                                color="#000"
                                disabled={false}
                                fill={false}
                                onClick={[Function]}
                                onComplete={[Function]}
                                overflow={false}
                                size={null}
                              >
                                <div
                                  aria-hidden={true}
                                  className="css-ugy6xl"
                                  data-test-id="Ripple"
                                  onClick={[Function]}
                                  style={null}
                                >
                                  <Magnifier>
                                    <Icon
                                      className=""
                                      color={null}
                                      content="<path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"/>"
                                      size="inherit"
                                      viewBox="0 0 24 24"
                                    >
                                      <svg
                                        className="css-yys9hb "
                                        dangerouslySetInnerHTML={
                                          Object {
                                            "__html": "<path d=\\"M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19l-4.99-5zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14z\\"/>",
                                          }
                                        }
                                        style={
                                          Object {
                                            "fill": null,
                                            "fontSize": "inherit",
                                          }
                                        }
                                        viewBox="0 0 24 24"
                                        xmlns="http://www.w3.org/2000/svg"
                                      />
                                    </Icon>
                                  </Magnifier>
                                </div>
                              </Ripple>
                            </button>
                          </SearchButton>
                        </Consumer>
                      </_default>
                      <Portal
                        name="nav-bar.navigator.icons.cart-button.before"
                        props={null}
                      />
                      <Portal
                        name="nav-bar.navigator.icons.cart-button"
                        props={
                          Object {
                            "CartButton": [Function],
                          }
                        }
                      >
                        <Component>
                          <div>
                             
                          </div>
                        </Component>
                      </Portal>
                      <Portal
                        name="nav-bar.navigator.icons.cart-button.after"
                        props={null}
                      />
                    </Portal>
                    <Portal
                      name="nav-bar.navigator.icons.after"
                      props={null}
                    />
                  </section>
                </Grid>
              </Portal>
              <Portal
                name="nav-bar.navigator.after"
                props={null}
              />
              <Portal
                name="nav-bar.progress-bar.before"
                props={null}
              />
              <Portal
                name="nav-bar.progress-bar"
                props={
                  Object {
                    "ProgressBar": [Function],
                  }
                }
              >
                <ProgressBar
                  isVisible={false}
                >
                  <Transition
                    appear={false}
                    enter={true}
                    exit={true}
                    in={false}
                    mountOnEnter={false}
                    onEnter={[Function]}
                    onEntered={[Function]}
                    onEntering={[Function]}
                    onExit={[Function]}
                    onExited={[Function]}
                    onExiting={[Function]}
                    timeout={300}
                    unmountOnExit={false}
                  >
                    <div
                      className="css-2m2907"
                      style={
                        Object {
                          "transform": "scale(1, 0)",
                        }
                      }
                    >
                      <div
                        className="css-aywzco"
                      />
                    </div>
                  </Transition>
                </ProgressBar>
              </Portal>
              <Portal
                name="nav-bar.progress-bar.after"
                props={null}
              />
            </header>
          </Portal>
          <Portal
            name="nav-bar.after"
            props={null}
          />
        </div>
      </Provider>
    </Navigator>
  </Connect(Navigator)>
</Provider>
`;
